### ISSUE TITLE
Nodeboxes with type=connected don't connect properly with the selflike

### ISSUE THREAD

**Issue by RichardTry** [minetest/minetest#10315]
##### Minetest version
```
Minetest 5.3.0 (Linux)
Using Irrlicht 1.8.4
BUILD_TYPE=Release
RUN_IN_PLACE=0
USE_GETTEXT=1
USE_SOUND=1
USE_CURL=1
USE_FREETYPE=1
USE_LUAJIT=1
STATIC_SHAREDIR="/usr/share/games/minetest"
STATIC_LOCALEDIR="/usr/share/locale"
```

##### OS / Hardware
Operating system: Debian 10
CPU: AMD Ryzen 3 2200G with Radeon Vega Graphics

##### Summary
Nodeboxes with type=connected do not connect to node from its "connects_to" list, if this node is also connected type nodebox and do not have "back connection".
Example from image:
Node1 is connection type nodebox which connects to default:stone and itself;
Node2 connects_to list is Node1 and itself, but as we see, there is no Node1 - Node2 connection;
Node2 will connect just in case of cross connection (If Node1 "connects back" to Node2).

![photo_2020-08-23_22-24-18](https://user-images.githubusercontent.com/16964020/90986926-683af600-e58f-11ea-9d51-cd8d4db8c2d3.jpg)

##### Steps to reproduce
Run this code as init.lua and make sure.

```
local size1 = 0.5 / 2
local size2 = 0.5 / 8

minetest.register_node("connections_bug:node1", {
	description = "Node 1",
	tiles = {
		"default_stone.png"
	},
	drawtype = "nodebox",
		node_box = {
		type = "connected",
		fixed          = {-size1, -size1, -size1, size1,  size1, size1},
		connect_top    = {-size1, -size1, -size1, size1,  0.5,  size1}, -- y+
		connect_bottom = {-size1, -0.5,  -size1, size1,  size1, size1}, -- y-
		connect_front  = {-size1, -size1, -0.5,  size1,  size1, size1}, -- z-
		connect_back   = {-size1, -size1,  size1, size1,  size1, 0.5 }, -- z+
		connect_left   = {-0.5,  -size1, -size1, size1,  size1, size1}, -- x-
		connect_right  = {-size1, -size1, -size1, 0.5,   size1, size1}, -- x+
		},
		connects_to = {"default:stone", "connections_bug:node1"},
	paramtype = "light",
	groups = {oddly_breakable_by_hand=2},
})

minetest.register_node("connections_bug:node2", {
	description = "Node 2",
	tiles = {
		"default_stone.png"
	},
	drawtype = "nodebox",
		node_box = {
		type = "connected",
		fixed          = {-size2, -size2, -size2, size2,  size2, size2},
		connect_top    = {-size2, -size2, -size2, size2,  0.5,  size2}, -- y+
		connect_bottom = {-size2, -0.5,  -size2, size2,  size2, size2}, -- y-
		connect_front  = {-size2, -size2, -0.5,  size2,  size2, size2}, -- z-
		connect_back   = {-size2, -size2,  size2, size2,  size2, 0.5 }, -- z+
		connect_left   = {-0.5,  -size2, -size2, size2,  size2, size2}, -- x-
		connect_right  = {-size2, -size2, -size2, 0.5,   size2, size2}, -- x+
		},
		connects_to = {"connections_bug:node1", "connections_bug:node2"},
	paramtype = "light",
	groups = {oddly_breakable_by_hand=2},
})
```

**Comment by paramat** [comment]
> Nodeboxes with type=connected do not connect to node from its "connects_to" list, if this node is also connected type nodebox and do not have "back connection".

Hmm, this seems possibly acceptable. But then, the request, to connect if either define a connection to the other, also seems acceptable.
Which is better behaviour and why?
If we change the behaviour, some mods might be relying on the current behaviour (not a blocker but needs considering if there is little advantage either way).

@sofar any thoughts?

**Comment by RichardTry** [comment]
> But then, the request, to connect if either define a connection to the other, also seems possibly acceptable.

Explain in more detail please, connection when requesting only from one side is just what I meant, isn't it?

Anyway, how can current behavior be better, if the connects_to nodes is clearly written in the node definition?

**Comment by paramat** [comment]
Well ...

Consider you have connected nodeboxes A and B.
A defines connection to B, but B does not define connection to A.
In this situation, you could argue that 'they should connect because A defines connection to B'.
But you can also argue it from B's perspective:
B does not define connection to A, so it makes sense they should not connect.

So both behaviours have good arguments.

This issue is arguing only from the perspective of A.

My initial opinion is that if either nodebox disallows a connection to the other then no connection should happen, both should agree on a connection.

Is the current behaviour actually a problem and why?

**Comment by RichardTry** [comment]
I don't want they both to connect each other, in your example, I just want A connect to B, B remains "fixed".

![screenshot_20200824_001012](https://user-images.githubusercontent.com/16964020/90988895-fff41080-e59e-11ea-9776-3238b21bb7e9.png)

There is no A or B perspective here.

**Comment by paramat** [comment]
Ahh i see.

**Comment by RichardTry** [comment]
In my mod, for example, connection nodeboxes are go beyond the node and go deeper into another nodebox, so it does not look messy. Screenshot was just for example.
In this logic, wood also do not agree to connects with the fence, but fence still does it.
We cannot know the shape of the nodebox to tell is it messy or not, so I think nodes should strictly follow connects_to list.

**Comment by paramat** [comment]
I accept that some users will use connected nodeboxes in a non-standard way, such that what you request would be beneficial for them. But we should consider majority usage. Usually, connections go halfway.
> In this logic, wood also do not agree to connects with the fence, but fence still does it.

Yes but it is probably intentional that normal nodes will agree to a connection to any connected nodebox that requests it. It seems they have to because they have no 'connects_to' field, so the only other option would be to disagree to all connections, which would remove the abilty for connected nodeboxes to connect to normal nodes.
> I think nodes should strictly follow connects_to list.

I think the 'connects_to' field does not mean 'draw this side of the connection in any situation'. I think 'connects_to' implies the entire connection, both sides of the connection.
I think the current behaviour is intentional and not a bug. Sofar tends to think features through carefully and connecting behaviour is a primary aspect he would have thought about.
I think the current behaviour is probably right for most uses.
I disapprove changing the behaviour in a non-optional way. Any alternative behaviour will have to be optional and not the default.

**Comment by RichardTry** [comment]
If the type of nodebox is "fixed" instead of "connected" it will look the same messy. So why cant node connect to another not-agree connection node, if it will look the same?
You can say, "there is no connects_to list", but if it will, the node does not need to know about the existence of the second node and "do not agree" to connect.

For example in my mod there is some hierarchy: Node1 (large branch) connects to tree, Node2 connects to Node1, etc.
Even if Node1 also have connects_to list, it is list of another group of major nodes, so from Node2 perspective, connecting to Node1 is the same as connecting to "connects_to"-less wood.

I think it is worth asking the opinion of @sofar and others.

**Comment by hecktest** [comment]
This is unsolveable actually. How do you know the B node is a centered box, and not a floor slab that happens to match?

**Comment by rubenwardy** [comment]
> I think the 'connects_to' field does not mean 'draw this side of the connection in any situation'. I think 'connects_to' implies the entire connection, both sides of the connection.

This is not true. You can make wooden fences connect to wood using connect_to

**Comment by paramat** [comment]
Adding this as a non-default optional behaviour seems fine.
It cannot be the default because changing the default behaviour would break mods relying on the current default behaviour.
> Although I think if some moder really comes across this, he will think this is a bug, as I did,

We have had connected nodeboxes for years and you are the first to consider the behaviour a 'bug'.
It seems you have a non-standard way of using connected nodeboxes that requires a different behaviour, that does not make it a 'bug', it is just problematic for your usage.
I might be wrong but i doubt there is any bug here. Both current and requested behaviours seem valid.

rubenwardy,
> This is not true. You can make wooden fences connect to wood using connect_to

Yes, but that is a special case i addressed above:
> it is probably intentional that normal nodes will agree to a connection to any connected nodebox that requests it. It seems they have to because they have no 'connects_to' field, so the only other option would be to disagree to all connections, which would remove the abilty for connected nodeboxes to connect to normal nodes.

... i was referring to the usual case of 2 connected nodeboxes.

**Comment by numberZero** [comment]
Sounds reasonable but that should be well documented. I did notice that while developing #6716 (obsolete):
![(dis)connected nodeboxes](https://user-images.githubusercontent.com/7352626/33911018-dcf514b8-dfa1-11e7-8d0a-f021864b03f3.png)

**Comment by RichardTry** [comment]
> Sounds reasonable but that should be well documented. I did notice that while developing #6716 (obsolete):

On your screenshot, on the contrary, everything looks correct. One node is connecting, another is not.

**Comment by Skamiz** [comment]
Today I wanted to make some lanterns which automatically attach to the bottom node if it has the group which I add to all nodes with standard cube collision. Which worked just fine as you can see:
![Screenshot from 2021-09-27 13-41-39](https://user-images.githubusercontent.com/48406239/134901663-18522e8d-8df9-4a7a-a68c-44936e654f01.png)
Imagine my surprise when I tell it to attach to a fence and it refuses to attach:
![Screenshot from 2021-09-27 13-48-20](https://user-images.githubusercontent.com/48406239/134902276-39f55946-2d1d-4844-aa75-15ac8ae1f0c8.png)
So I do some experimenting and find out that it is caused by "connected" nodebox. A bit later I find this issue and am baffled.
,
1) Why the heck should the fence get any decision power over what's placed on top of it? Or, worded differently, why am I apparently required to change the fence definition to account for the lantern when the fence is a complete non-actor in this. If this were a mod for default minetest game, am I supposed to change the fence definition just so the lantern can sit on them properly?

2) If I didn't want my lantern to connect to the fence, **I wouldn't tell it to connect to the fence!**


No node boxes are crossing the boundary of their node, so you don't get to claim this is non-standard usage. This is a practical, real example where the current behavior is incorrect.

I would dearly love to see a practical example where the ability to disallow another node to make it's connection is useful. I would also like to see an example of someone actually using it, instead of this "hypothetically someone could have". And lastly, as I saw mentioned at a few points in the past, since the current behavior isn't documented, it's not part of the official API and nobody should depend on it anyway.

In conclusion, until and unless someone brings **practical** examples of the current behavior being good for something, I consider it a bug.

**Comment by The4codeblocks** [comment]
> In conclusion, until and unless someone brings **practical** examples of the current behavior being good for something, I consider it a bug.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/91603f6f-45ac-42ff-8081-fab1eb4f5575" />

tubes only connecting if both tubes agree
